### DISCUSSION TITLE
General Welded Cross-Section in GH

### DISCUSSION THREAD

**Discussion by ktrn4896** [idea-statica/ideastatica-public#506] [Rhino/Grasshopper] [answered]
Hi! 

I am trying to create a general welded cross section shape in a custom C# component. 
So far I just copied the example code from here [Adding Cross-Section Properties to IOM Open Model](https://developer.ideastatica.com/docs/iom/iom_cross_sections.html#creating-a-general-welded-cross-section) and it seems to work (see [generic_cs.zip](https://github.com/idea-statica/ideastatica-public/files/15139453/generic_cs.zip)). 
But to use the created cross section shape in the "Create Cross Section (IOM)" component, I need to wrap it. 
I tried the method, suggested to me by @nathanlukeIDEA here: [Component for Grasshopper #447](https://github.com/idea-statica/ideastatica-public/discussions/447#discussioncomment-8511347), namely: 
   `CrossSectionComponent componentCrossSection = new CrossSectionComponent() { Id = 1, Name = "1" };`
   `...`
   `RhinoCrossSectionShape rh_cs = new RhinoCrossSectionShape(componentCrossSection);`
   `GH_IdeaCrossSectionShape gh_cs = new  GH_IdeaCrossSectionShape(rh_cs);`
    `cs_shape = gh_cs; // output` 
But I got the "Error (CS0144): **Cannot create an instance of the abstract class or interface 'IdeaStatiCa.IOM2Rhino.RhinoCrossSectionShape**' (line 103)".
Is it possible to create a generic cross section in GH by code at the moment? If yes, could you, please, give me a "wrapping" example. 

Best regards, 
Ekaterina

**Comment by nathanlukeIDEA** [comment, marked as answer]
Hi @ktrn4896,

[Create Connection using IOM Components_custom_css_test.zip](https://github.com/idea-statica/ideastatica-public/files/15231737/Create.Connection.using.IOM.Components_custom_css_test.zip)

Thanks for your patience. Please find attached and below and example. Again you will need to manually reference all the .dlls. 

The general cross-section is based on the one here: https://developer.ideastatica.com/docs/iom/iom_cross_sections.html#creating-a-general-welded-cross-section

```csharp
using IdeaRS.OpenModel;
using IdeaRS.OpenModel.CrossSection;
using IdeaRS.OpenModel.Material;
using IdeaRS.OpenModel.Geometry2D;
using IdeaRS.OpenModel.Geometry3D;
using IdeaStatiCa.Grasshopper;
using IdeaStatiCa.IOM2Rhino;

  private void RunScript(object x, object y, ref object A)
  {


    //Create Material

    MatSteelEc2 matS = new MatSteelEc2();
    matS.Name = "S275";
    matS.Id = 1;
    matS.UnitMass = 7850.0;
    matS.E = 200e9;
    matS.Poisson = 0.2;
    matS.G = 83.333e9;
    matS.SpecificHeat = 0.6;
    matS.ThermalExpansion = 0.00001;
    matS.ThermalConductivity = 45;
    matS.fy = 235e6;
    matS.fu = 360e6;
    matS.fy40 = 215e6;
    matS.fu40 = 340e6;

    matS.StateOfThermalConductivity = ThermalConductivityState.Code;
    matS.StateOfThermalExpansion = ThermalExpansionState.Code;
    matS.StateOfThermalSpecificHeat = ThermalSpecificHeatState.Code;
    matS.StateOfThermalStressStrain = ThermalStressStrainState.Code;

    matS.DiagramType = SteelDiagramType.Bilinear;

    //A = matS;

    RhinoMaterial rhMaterial = new RhinoMaterial(matS);


    //CREATE CROSS-SECTION

    CrossSectionComponent componentCrossSection = new CrossSectionComponent() { Id = 1, Name = "myComponentCrossSection" };

    //Component 1
    CssComponent comp1 = new CssComponent() { Id = 1, Material = new ReferenceElement(matS), Phase = 0 };

    Region2D region = new Region2D();
    PolyLine2D outline = new PolyLine2D();

    outline.StartPoint = new Point2D() { X = 0, Y = 0 };
    outline.Segments.Add(new LineSegment2D() { EndPoint = new Point2D() { X = 0.2, Y = 0 } });
    outline.Segments.Add(new LineSegment2D() { EndPoint = new Point2D() { X = 0.2, Y = -0.015 } });
    outline.Segments.Add(new LineSegment2D() { EndPoint = new Point2D() { X = 0, Y = -0.015 } });
    outline.Segments.Add(new LineSegment2D() { EndPoint = new Point2D() { X = 0, Y = 0 } });

    region.Outline = outline;

    comp1.Geometry = region;
    componentCrossSection.Components.Add(comp1);

    //Component2
    CssComponent comp2 = new CssComponent() { Id = 2, Material = new ReferenceElement(matS), Phase = 0 };
    region = new Region2D();
    outline = new PolyLine2D();

    outline.StartPoint = new Point2D() { X = 0.095, Y = -0.015 };
    outline.Segments.Add(new LineSegment2D() { EndPoint = new Point2D() { X = 0.105, Y = -0.015 } });
    outline.Segments.Add(new LineSegment2D() { EndPoint = new Point2D() { X = 0.105, Y = -0.14 } });
    outline.Segments.Add(new LineSegment2D() { EndPoint = new Point2D() { X = 0.095, Y = -0.14 } });
    outline.Segments.Add(new LineSegment2D() { EndPoint = new Point2D() { X = 0.095, Y = -0.015 } });
    region.Outline = outline;

    comp2.Geometry = region;
    componentCrossSection.Components.Add(comp2);

    RhinoComponentCrossSection rhinoCrossSection = new RhinoComponentCrossSection(null, componentCrossSection, rhMaterial);

    GH_IdeaCrossSection gh_css = new GH_IdeaCrossSection(rhinoCrossSection);

    A = gh_css;

  }


```
